kubernetes/istio_rules_service.go

```diff
@@ -37,153 +39,97 @@ func (in *IstioClient) GetIstioRuleDetails(namespace string, istiorule string) (
-	for i := 0; i < len(instances.(instancesType)); i++ {
-		instanceResponse := <-instancesChan
-		if instanceResponse.err != nil {
-			return nil, instanceResponse.err
-		}
-		istioRuleAction.Instances = append(istioRuleAction.Instances, instanceResponse.result)
-	}
+func (in *IstioClient) GetAdapter(namespace, adapterType, adapterName string) (IstioObject, error) {
+	return in.getAdapterTemplate(namespace, "adapter", adapterType, adapterName, adapterPlurals)
+}
 
-	return &istioRuleAction, nil
+func (in *IstioClient) GetTemplate(namespace, templateType, templateName string) (IstioObject, error) {
+	return in.getAdapterTemplate(namespace, "template", templateType, templateName, templatePlurals)
 }
 
-func (in *IstioClient) getActionHandler(namespace string, handlerType string, handlerName string, handlerChan chan<- istioResponse) {
-	handlerTypePlural, ok := istioTypePlurals[handlerType]
-	if !ok {
-		log.Warningf("Handler type %s is not supported", handlerType)
-		handlerChan <- istioResponse{}
-		return
-	}
-	result, err := in.istioConfigApi.Get().Namespace(namespace).Resource(handlerTypePlural).SubResource(handlerName).Do().Get()
-	istioObject, ok := result.(IstioObject)
-	if !ok {
-		istioObject = nil
-		if err == nil {
-			err = fmt.Errorf("%s/%s doesn't return a valid IstioObject", handlerType, handlerName)
+func (in *IstioClient) getAdaptersTemplates(namespace string, itemType string, pluralsMap map[string]string) ([]IstioObject, error) {
+	resultsChan := make(chan istioResponse)
+	for name, plural := range pluralsMap {
+		go func(name, plural string) {
+			results, err := in.istioConfigApi.Get().Namespace(namespace).Resource(plural).Do().Get()
+			istioObjects := istioResponse{}
+			resultList, ok := results.(*GenericIstioObjectList)
+			if !ok {
+				istioObjects.results = nil
+				istioObjects.err = fmt.Errorf("%s doesn't return a %s list", namespace, plural)
+			}
+			if err == nil {
+				istioObjects.results = make([]IstioObject, 0)
+				for _, result := range resultList.Items {
+					adapter := result.DeepCopyIstioObject()
+					// We need to specifically add the adapter name in the label
+					if adapter.GetObjectMeta().Labels == nil {
+						objectMeta := adapter.GetObjectMeta()
+						objectMeta.Labels = make(map[string]string)
+						adapter.SetObjectMeta(objectMeta)
+					}
+					adapter.GetObjectMeta().Labels[itemType] = name
+					// To support plural, we have only adapter/template -> adapters/templates
+					adapter.GetObjectMeta().Labels[itemType] = name
+					adapter.GetObjectMeta().Labels[itemType + "s"] = plural
+					istioObjects.results = append(istioObjects.results, adapter)
+					istioObjects.err = nil
+				}
+			} else {
+				istioObjects.results = nil
+				istioObjects.err = err
+			}
+			resultsChan <- istioObjects
+		}(name, plural)
+	}
+
+	results := make([]IstioObject, 0)
+	for i := 0; i < len(pluralsMap); i++ {
+		adapter := <-resultsChan
+		if adapter.err != nil {
+			return nil, adapter.err
```

Comment:
Just a note here: I think it's ok to leave without closing the channel as [it will be garbage-collected](https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open).
However maybe we should want to continue collecting responses before returning immediately? (and store the error in a variable). That's particularly interesting in the case of read-permission denied on some objects: we would still continue to display objects where we have read permissions granted, rather than failing for everything. (That's a pattern that we don't apply in many parts of the code but I think it should be nice to start thinking like that)

Reply:
Yes, in this KIALI-1947 I planned to work only on the split of rules/templates/adapters and consolidate it from the backend to the ui.
But you are right, in KIALI-1654 I wanted to link this logic with the permissions one, just to separate them.
But yes, it makes sense, is something to take into consideration.